examples/pennylane_run_doubly_stochastic.py

```diff
@@ -0,0 +1,312 @@
+opt = qml.GradientDescentOptimizer(eta)
+
+for _ in range(steps):
+    cost_SGD100.append(qnode_SGD(params_SGD100))
+    params_SGD100 = opt.step(qnode_SGD, params_SGD100)
+
+
+##############################################################################
+# We can now plot the cost against optimization step for the three cases.
+
+from matplotlib import pyplot as plt
+
+plt.style.use("seaborn")
+plt.plot(cost_GD[:100], label="Vanilla gradient descent")
+plt.plot(cost_SGD100[:100], label="Stochastic gradient descent, $shots=100$")
+plt.plot(cost_SGD1[:100], ".", label="Stochastic gradient descent, $shots=1$")
+
+# analytic ground state
+min_energy = min(np.linalg.eigvalsh(H))
+plt.hlines(min_energy, 0, 100, linestyles=":", label="Ground-state energy")
+
+plt.ylabel("Cost function value")
+plt.xlabel("Optimization steps")
+plt.legend()
+plt.show()
+
+##############################################################################
+# Using the trained parameters from each optimization strategy, we can
+# evaluate the analytic quantum device:
+
+print("Vanilla gradient descent min energy = ", qnode_GD(params_GD))
+print("Stochastic gradient descent (shots=100) min energy = ", qnode_GD(params_SGD100))
+print("Stochastic gradient descent (shots=1) min energy = ", qnode_GD(params_SGD1))
+
+##############################################################################
+# Amazingly, we see that even the ``shots=1`` optimization converged
+# to a reasonably close approximation of ground-state energy!
+
+##############################################################################
+# Doubly stochastic gradient descent for VQE
+# ------------------------------------------
+#
+# As noted in `Sweke et al. (2019) <https://arxiv.org/abs/1910.01155>`__,
+# variational quantum algorithms often include terms consisting of linear combinations
+# of expectation values. This is true of the parameter-shift rule (where the
+# gradient of each parameter is determined by shifting the parameter by macroscopic
+# amounts and taking the difference), as well as VQE, where the Hamiltonian
+# is usually decomposed into a sum of Pauli expectation values.
+#
+# Consider the Hamiltonian from the previous section. As this Hamiltonian is an
+# Hermitian observable, we can always express it as a sum of Pauli matrices using
+# the relation
+#
+# .. math::
+#
+#     H = \sum_{i,j=I,x,y,z} a_{i,j} (\sigma_i\otimes \sigma_j),
+#         ~~a_{i,j} = \frac{1}{4}\text{tr}[(\sigma_i\otimes \sigma_j )H].
+#
+# Applying this, we can see that
+#
+# .. math::
+#
+#     H = 4  + 2I\otimes X + 4I \otimes Z - X\otimes X + 5 Y\otimes Y + 2Z\otimes X.
+#
+# To perform "doubly stochastic" gradient descent, we simply apply the stochastic
+# gradient descent approach from above, but in addition also uniformly sample
+# a subset of the terms for the Hamiltonian expectation at each optimization step.
+# This inserts another element of stochasticity into the system---all the while

```

Comment:
```suggestion
# This inserts another element of stochasticity into the system --- all the while
```
?

Reply:
Depends on the convention :slightly_smiling_face: Some conventions use---like this, while in others --- this is okay. Tends to be an American/British distinction as well.